src/puppetlabs/jdbc_util/core.clj

```diff
@@ -159,3 +160,17 @@
                        [sql 0]
                        seq-params-w-indices)]
     (vec (conj (flatten parameters) sql'))))
+
+(defn wrap-with-db-permission-error-handling
+  "A ring middleware to catch and report postgresql permission-denied errors."
+  [handler]
+  (fn [request]
+    (try
+      (handler request)
+      (catch PSQLException e
+        (if (= "42501" (.getSQLState e))
+          {:status 403
+           :headers {"Content-Type" "application/json"}
+           :body (json/generate-string {:kind "db-permission-error"
+                                        :msg (tru "Insufficient database permissions")})}
```

Comment:
As long as we have an end result where we have a full sentence, that's fine, but that full sentence needs to exist somewhere.